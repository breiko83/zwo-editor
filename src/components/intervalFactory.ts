import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
