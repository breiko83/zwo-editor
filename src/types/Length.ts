export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
