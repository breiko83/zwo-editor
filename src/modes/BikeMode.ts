import { LengthType } from "../types/LengthType";
import { Duration, Length } from "../types/Length";
import { SportType } from "../types/SportType";
import Mode from "./Mode";
import { Interval } from "../types/Interval";

export default class BikeMode extends Mode {
  private ftp: number;
  private weight: number;
  public readonly sportType: SportType = "bike";
  public readonly lengthType: LengthType = "time";

  constructor(ftp: number, weight: number) {
    super();
    this.ftp = ftp;
    this.weight = weight;
  }

  power(intensity: number): number {
    return intensity * this.ftp;
  }

  wkg(intensity: number): number {
    return this.power(intensity) / this.weight;
  }

  duration(length: Length): Duration {
    if (length instanceof Duration) {
      return length;
    } else {
      throw new Error("Unexpected length:Distance encountered in BikeMode");
    }
  }

  intervalDuration(interval: Interval): Duration {
    switch (interval.type) {
      case 'free':
      case 'steady':
      case 'ramp':
        return this.duration(interval.length);
      case 'repetition': {
        const onDuration = this.duration(interval.onLength);
        const offDuration = this.duration(interval.offLength);
        return new Duration(interval.repeat * (onDuration.seconds + offDuration.seconds));
      }
    }
  }
}
