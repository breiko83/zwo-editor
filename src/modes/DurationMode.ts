import { distanceMultiplier, durationMultiplier, intensityMultiplier } from "../components/Bar/multipliers";
import { Duration, Length } from "../types/Length";
import { floor } from "../utils/math";

// Base class for duration-based Mode classes
export default abstract class DurationMode {
  percentage(intensity: number): number {
    return Math.round(intensity * 100);
  }

  lengthToWidth(length: Length): number {
    if (length instanceof Duration) {
      return length.seconds / durationMultiplier;
    } else {
      return length.meters / distanceMultiplier;
    }
  }

  widthToLength(width: number): Duration {
    return new Duration(floor(width * durationMultiplier, 5));
  }

  intensityToHeight(intensity: number): number {
    return intensity * intensityMultiplier;
  }

  heightToIntensity(height: number): number {
    return height / intensityMultiplier;
  }

  duration(length: Length): Duration {
    if (length instanceof Duration) {
      return length;
    } else {
      return new Duration(0); // TODO
    }
  }
}
