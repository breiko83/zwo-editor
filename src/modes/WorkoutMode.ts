import BikeMode from "./BikeMode";
import RunMode from "./RunMode";

export type WorkoutMode = BikeMode | RunMode;
