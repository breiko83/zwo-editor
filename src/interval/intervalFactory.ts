import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
