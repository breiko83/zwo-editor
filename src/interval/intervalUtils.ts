import { Zones } from "../types/Zones";
import { Interval } from "../types/Interval";
import { Duration } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";
import { move, propEq, update } from "ramda";

// Helpers for transforming intervals array

function updateById(id: string, transform: (interval: Interval) => Interval, intervals: Interval[]): Interval[] {
  const index = intervals.findIndex(propEq('id', id));
  if (index === -1) {
    return intervals;
  }

  return update(index, transform(intervals[index]), intervals);
}

export function updateIntervalDuration(id: string, dDuration: Duration, intervals: Interval[], mode: WorkoutMode): Interval[] {
  return updateById(id, (interval) => {
    if (interval.type === 'steady') {
      const seconds = mode.intervalDuration(interval).seconds + dDuration.seconds;
      if (seconds > 0) {
        return { ...interval, length: new Duration(seconds) };
      }
    }
    return interval;
  }, intervals);
}

export function updateIntervalIntensity(id: string, dIntensity: number, intervals: Interval[]): Interval[] {
  return updateById(id, (interval) => {
    if (interval.type === 'steady') {
      const intensity = interval.intensity + dIntensity;
      if (intensity > Zones.Z1.min) {
        return { ...interval, intensity };
      }
    }
    return interval;
  }, intervals);
}

export function moveInterval(id: string, direction: -1 | 1, intervals: Interval[]): Interval[] {
  const oldIndex = intervals.findIndex(propEq('id', id));
  const newIndex = oldIndex + direction;

  if (oldIndex < 0 || newIndex < 0 || newIndex >= intervals.length) {
    return intervals;
  }

  return move(oldIndex, newIndex, intervals);
}
