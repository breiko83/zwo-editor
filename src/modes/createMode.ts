import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
