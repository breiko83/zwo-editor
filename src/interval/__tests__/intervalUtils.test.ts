import { moveInterval, updateIntervalDuration, updateIntervalIntensity } from '../intervalUtils';
import { Duration } from '../../types/Length';
import intervalFactory from '../intervalFactory';
import BikeMode from '../../modes/BikeMode';
import { WorkoutMode } from '../../modes/WorkoutMode';

const defaultBikeMode = () => new BikeMode(200, 75);

function deepFreeze<T>(object: T): T {
  for (const name of Object.getOwnPropertyNames(object)) {
    const value = (object as any)[name];
    if (value && typeof value === "object") {
      deepFreeze(value);
    }
  }
  return Object.freeze(object);
}

// Use freeze to ensure we aren't accidentally mutating our interval list
const defaultIntervals = (mode: WorkoutMode) => deepFreeze([
  intervalFactory.steady({
    id: "#1",
    length: new Duration(60),
    intensity: 0.5
  }, mode),
  intervalFactory.steady({
    id: "#2",
    length: new Duration(120),
    intensity: 1.0
  }, mode),
  intervalFactory.steady({
    id: "#3",
    length: new Duration(30),
    intensity: 0.75
  }, mode),
]);

describe('intervalUtils', () => {
  describe('updateIntervalDuration', () => {
    it('increases or decreses duration of specified interval', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalDuration("#2", new Duration(10), intervals, mode)).toMatchSnapshot();
      expect(updateIntervalDuration("#3", new Duration(-10), intervals, mode)).toMatchSnapshot();
    });

    it('does not allow decreasing duration to zero or below it', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalDuration("#1", new Duration(-65), intervals, mode)).toEqual(intervals);
      expect(updateIntervalDuration("#1", new Duration(-60), intervals, mode)).toEqual(intervals);
    });

    it('does nothing when ID not found', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalDuration("#blah", new Duration(10), intervals, mode)).toEqual(intervals);
    });
  });

  describe('updateIntervalIntensity', () => {
    it('increases or decreses duration of specified interval', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#2", 0.05, intervals)).toMatchSnapshot();
      expect(updateIntervalIntensity("#3", -0.05, intervals)).toMatchSnapshot();
    });

    it('does not allow decreasing intensity to zero or below it', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#1", -0.55, intervals)).toEqual(intervals);
      expect(updateIntervalIntensity("#1", -0.5, intervals)).toEqual(intervals);
    });

    it('does nothing when ID not found', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#blah", 0.1, intervals)).toEqual(intervals);
    });
  });

  describe('moveInterval', () => {
    it('moves interval index by +1 or -1', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#2", -1, intervals)).toMatchSnapshot();
      expect(moveInterval("#2", +1, intervals)).toMatchSnapshot();
    });

    it('does not allow moving past start or end', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#1", -1, intervals)).toEqual(intervals);
      expect(moveInterval("#3", +1, intervals)).toEqual(intervals);
    });

    it('does nothing when ID not found', () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#blah", 1, intervals)).toEqual(intervals);
    });
  });
});
