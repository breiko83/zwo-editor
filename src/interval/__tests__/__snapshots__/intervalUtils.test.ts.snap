// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intervalUtils moveInterval moves interval index by +1 or -1 1`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1,
    "length": Duration {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.75,
    "length": Duration {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils moveInterval moves interval index by +1 or -1 2`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.75,
    "length": Duration {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1,
    "length": Duration {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalDuration increases or decreses duration of specified interval 1`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1,
    "length": Duration {
      "seconds": 130,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.75,
    "length": Duration {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalDuration increases or decreses duration of specified interval 2`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1,
    "length": Duration {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.75,
    "length": Duration {
      "seconds": 20,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalIntensity increases or decreses duration of specified interval 1`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1.05,
    "length": Duration {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.75,
    "length": Duration {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalIntensity increases or decreses duration of specified interval 2`] = `
Array [
  Object {
    "cadence": 0,
    "id": "#1",
    "intensity": 0.5,
    "length": Duration {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#2",
    "intensity": 1,
    "length": Duration {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": 0,
    "id": "#3",
    "intensity": 0.7,
    "length": Duration {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;
